File: Java/CesiumLanguageWriter/translatedSrc/cesiumlanguagewriter/advanced/CesiumInterpolatablePropertyWriter.java
package cesiumlanguagewriter.advanced;


import agi.foundation.compatibility.*;
import cesiumlanguagewriter.*;
import javax.annotation.Nonnull;

/**
 * A CZML writer for a property that represents a value that may be sampled
 and interpolated.  The property may be defined over a single interval or over multiple intervals.
 * @param <TDerived> The type of the class derived from this one.
 */
@SuppressWarnings( {
        "unused",
        "deprecation",
        "serial"
})
public abstract class CesiumInterpolatablePropertyWriter<TDerived extends CesiumInterpolatablePropertyWriter<TDerived>> extends CesiumPropertyWriter<TDerived> implements
        ICesiumInterpolatablePropertyWriter {
    /**
    * Initializes a new instance.
    * @param propertyName The name of the property.
    */
    protected CesiumInterpolatablePropertyWriter(@Nonnull String propertyName) {
        super(propertyName);
    }

    /**
    * Initializes a new instance as a copy of an existing instance.
    * @param existingInstance The existing instance to copy.
    */
    protected CesiumInterpolatablePropertyWriter(@Nonnull CesiumInterpolatablePropertyWriter<TDerived> existingInstance) {
        super(existingInstance);
    }

    /**
    * Writes the interpolation algorithm to use to interpolate the sampled data in this interval.
    * @param interpolationAlgorithm The interpolation algorithm.
    */
    public final void writeInterpolationAlgorithm(@Nonnull CesiumInterpolationAlgorithm interpolationAlgorithm) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("interpolationAlgorithm");
        getOutput().writeValue(CesiumFormattingHelper.interpolationAlgorithmToString(interpolationAlgorithm));
    }

    /**
    * Writes the degree of polynomial to use to interpolate sampled data in this interval.
    * @param degree The degree.
    */
    public final void writeInterpolationDegree(int degree) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("interpolationDegree");
        getOutput().writeValue(degree);
    }

    /**
    * Writes the type of extrapolation to perform when a value is requested at a time after any available samples.
    * @param extrapolationType The extrapolation type.
    */
    public final void writeForwardExtrapolationType(@Nonnull CesiumExtrapolationType extrapolationType) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("forwardExtrapolationType");
        getOutput().writeValue(CesiumFormattingHelper.extrapolationTypeToString(extrapolationType));
    }

    /**
    * Writes the amount of time to extrapolate forward before the property becomes undefined.  A value of 0 will extrapolate forever.
    * @param duration The duration.
    */
    public final void writeForwardExtrapolationDuration(@Nonnull Duration duration) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("forwardExtrapolationDuration");
        getOutput().writeValue(duration.getTotalSeconds());
    }

    /**
    * Writes the type of extrapolation to perform when a value is requested at a time before any available samples.
    * @param extrapolationType The extrapolation type.
    */
    public final void writeBackwardExtrapolationType(@Nonnull CesiumExtrapolationType extrapolationType) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("backwardExtrapolationType");
        getOutput().writeValue(CesiumFormattingHelper.extrapolationTypeToString(extrapolationType));
    }

    /**
    * Writes the amount of time to extrapolate backward before the property becomes undefined.  A value of 0 will extrapolate forever.
    * @param duration The duration.
    */
    public final void writeBackwardExtrapolationDuration(@Nonnull Duration duration) {
        openIntervalIfNecessary();
        getOutput().writePropertyName("backwardExtrapolationDuration");
        getOutput().writeValue(duration.getTotalSeconds());
    }
}

File: DotNet/CesiumLanguageWriter/Advanced/CesiumInterpolatablePropertyWriter.cs
namespace CesiumLanguageWriter.Advanced
{
    /// <summary>
    /// A <topic name="Cesium">Cesium</topic> writer for a property that represents a value that may be sampled
    /// and interpolated.  The property may be defined over a single interval or over multiple intervals.
    /// </summary>
    /// <typeparam name="TDerived">The type of the class derived from this one.</typeparam>
    public abstract class CesiumInterpolatablePropertyWriter<TDerived> : CesiumPropertyWriter<TDerived>, ICesiumInterpolationInformationWriter
        where TDerived : CesiumInterpolatablePropertyWriter<TDerived>
    {
        /// <summary>
        /// Initializes a new instance.
        /// </summary>
        /// <param name="propertyName">The name of the property.</param>
        public CesiumInterpolatablePropertyWriter(string propertyName)
            : base(propertyName)
        {
        }

        /// <summary>
        /// Initializes a new instance as a copy of an existing instance.
        /// </summary>
        /// <param name="existingInstance">The existing instance to copy.</param>
        protected CesiumInterpolatablePropertyWriter(CesiumInterpolatablePropertyWriter<TDerived> existingInstance)
            : base(existingInstance)
        {
        }

        /// <summary>
        /// Writes the interpolation algorithm to use to interpolate the sampled data in this interval.
        /// </summary>
        /// <param name="interpolationAlgorithm">The interpolation algorithm.</param>
        public void WriteInterpolationAlgorithm(CesiumInterpolationAlgorithm interpolationAlgorithm)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("interpolationAlgorithm");
            Output.WriteValue(CesiumFormattingHelper.InterpolationAlgorithmToString(interpolationAlgorithm));
        }

        /// <summary>
        /// Writes the degree of polynomial to use to interpolate sampled data in this interval.
        /// </summary>
        /// <param name="degree">The degree.</param>
        public void WriteInterpolationDegree(int degree)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("interpolationDegree");
            Output.WriteValue(degree);
        }

        /// <summary>
        /// Writes the type of extrapolation to perform when a value is requested at a time after any available samples.
        /// </summary>
        /// <param name="extrapolationType">The extrapolation type.</param>
        public void WriteForwardExtrapolationType(CesiumExtrapolationType extrapolationType)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("forwardExtrapolationType");
            Output.WriteValue(CesiumFormattingHelper.ExtrapolationTypeToString(extrapolationType));
        }

        /// <summary>
        /// Writes the amount of time to extrapolate forward before the property becomes undefined.  A value of 0 will extrapolate forever.
        /// </summary>
        /// <param name="duration">The duration.</param>
        public void WriteForwardExtrapolationDuration(Duration duration)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("forwardExtrapolationDuration");
            Output.WriteValue(duration.TotalSeconds);
        }

        /// <summary>
        /// Writes the type of extrapolation to perform when a value is requested at a time before any available samples.
        /// </summary>
        /// <param name="extrapolationType">The extrapolation type.</param>
        public void WriteBackwardExtrapolationType(CesiumExtrapolationType extrapolationType)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("backwardExtrapolationType");
            Output.WriteValue(CesiumFormattingHelper.ExtrapolationTypeToString(extrapolationType));
        }

        /// <summary>
        /// Writes the amount of time to extrapolate backward before the property becomes undefined.  A value of 0 will extrapolate forever.
        /// </summary>
        /// <param name="duration">The duration.</param>
        public void WriteBackwardExtrapolationDuration(Duration duration)
        {
            OpenIntervalIfNecessary();

            Output.WritePropertyName("backwardExtrapolationDuration");
            Output.WriteValue(duration.TotalSeconds);
        }
    }
}